### ISSUE TITLE
Shoule NIX_ENFORCE_PURITY really silently drop ld arguments?

### ISSUE THREAD

**Issue by nh2** [NixOS/nixpkgs#24660]
I found that in https://github.com/NixOS/nixpkgs/blob/225602d9f2fc5d437e4996c9697dd272d18846c0/pkgs/build-support/cc-wrapper/ld-wrapper.sh#L18 there is a wrapper around `ld` that silently drops paths given to `ld` if `NIX_ENFORCE_PURITY` is set, if those paths are outside of the nix store path.

That doesn't look like a good idea to me, it is really hard to debug when that happens and your link just fails and the only thing you do is `strace` all the way down to see what paths are missing.

Should it not simply fail instead, complaining about arguments that are outside the store path (and telling you which ones)?

**Comment by copumpkin** [comment]
Yeah, I'd like that. Would help us catch and patch more packages that try to do impure things. If we were better about signal-to-noise ratio we could also emit a loud warning, but our builds tend to be so noisy that I doubt anyone would notice 😄

**Comment by mmahut** [comment]
Any news on this issue?

**Comment by nh2** [comment]
still important to me

**Comment by nh2** [comment]
still important to me

**Comment by bergkvist** [comment]
This is also relevant for the gcc wrapper.

A loud warning would definitely be very useful here. Not because everyone would see it - but because you would find it if you are looking for the reason your build failed. And that also shouldn't break anything.

I was trying to use gcc in an impure way inside the shellHook of pkgs.mkShell, and spent an hour confused about why the compiler couldn't find the include files I tried to pass it.

Turns out `NIX_ENFORCE_PURITY` was set during the invocation of shellHook, but unset directly after the invocation of shellHook, meaning compilation worked as long as I ran it manually inside nix-shell, but not if I put it inside of the shellHook.
